File: heimbohrtechnik/templates/pages/schlammanlieferung.js
$('document').ready(function(){
    make();
    run();
});


function make() {
    // reset values
    document.getElementById('full_weight').value = null;
    // rebrand
    var brands = document.getElementsByClassName('navbar-brand');
    for (var e = 0; e < brands.length; e++) {
        brands[e].innerHTML = "<span>MudEX AG</span>";
    }
    // attach button handlers
    $(".btn-weigh").on('click', function() { 
        var full_weight = document.getElementById('full_weight').value;
        var target = 'full_weight';
        if ((full_weight) && (full_weight > 0)) {
            target = 'empty_weight';
        }
        frappe.call({
            'method': 'heimbohrtechnik.mudex.doctype.truck_scale.truck_scale.get_weight',
            'args': {
                'truck_scale': document.getElementById('scale').value
            },
            'callback': function(r) {
                if (typeof r.message !== 'undefined') {
                    document.getElementById(target).value = r.message.weight;
                    if (target == "full_weight") {
                        document.getElementById('full_time').value = get_now();
                        document.getElementById('full_process_id').value = r.message.process_id;
                    } else {
                        document.getElementById('empty_time').value = get_now();
                        document.getElementById('empty_process_id').value = r.message.process_id;
                    }
                    compute_net();
                } else {
                    console.log("Invalid response");
                }
            }
        });
    });
    $(".btn-submit").on('click', function() { 
        frappe.call({
            'method': 'heimbohrtechnik.templates.pages.schlammanlieferung.insert_delivery',
            'args': {
                'truck': document.getElementById('truck').value, 
                'customer': document.getElementById('customer').value, 
                'object': document.getElementById('object').value, 
                'full_weight': document.getElementById('full_weight').value, 
                'empty_weight': document.getElementById('empty_weight').value, 
                'net_weight': document.getElementById('net_weight').value, 
                'traces': [{
                    'date': document.getElementById('full_time').value,
                    'scale': document.getElementById('scale').value,
                    'weight': document.getElementById('full_weight').value,
                    'process_id': document.getElementById('full_process_id').value
                },
                {
                    'date': document.getElementById('empty_time').value,
                    'scale': document.getElementById('empty_scale').value,
                    'weight': document.getElementById('empty_weight').value,
                    'process_id': document.getElementById('empty_process_id').value
                }],
                'load_type': document.getElementById('load_type').value,
                'ph': document.getElementById('ph').value
            },
            'callback': function(r) {
                if (typeof r.message !== 'undefined') {
                    var btn_submit = document.getElementById('submit');
                    btn_submit.disabled = true;
                    btn_submit.style.visibility = "hidden";
                    var row_success = document.getElementById('finalised-row');
                    row_success.style.visibility = "visible";
                    var txt_success = document.getElementById('result-doc');
                    txt_success.innerHTML = r.message.delivery;
                    document.getElementById('load_type').disabled = true;
                } else {
                    console.log("Invalid response");
                }
            }
        });
        
    });
}
function run() {
    // get command line parameters
    var arguments = window.location.toString().split("?");
    if (!arguments[arguments.length - 1].startsWith("http")) {
        var args_raw = arguments[arguments.length - 1].split("&");
        var args = {};
        args_raw.forEach(function (arg) {
            var kv = arg.split("=");
            if (kv.length > 1) {
                args[kv[0]] = kv[1];
            }
        });
        if (args['truck']) {
            document.getElementById('truck').value = args['truck'];
            get_truck_weight(args['truck']);
        }
        if (args['customer']) {
            document.getElementById('customer').value = args['customer'];
        }
        if (args['key']) {
            document.getElementById('key').value = args['key'];
        }
        if (args['object']) {
            var object_field = document.getElementById('object');
            object_field.value = args['object'];
            object_field.setAttribute('readonly', true);
            // fetch object data
            get_object_details(args['object'], args['truck'], args['customer'], args['key']);
        }
        
    } else {
        // no arguments provided
        
    }
    // get scale 
    frappe.call({
        'method': 'heimbohrtechnik.templates.pages.schlammanlieferung.get_default_scale',
        'callback': function(r) {
            document.getElementById('scale').value  = r.message
        }
    });
    // get load types
    frappe.call({
        'method': 'heimbohrtechnik.templates.pages.schlammanlieferung.get_load_types',
        'args': {
            'object': args['object'] || null
        },
        'callback': function(r) {
            console.log(r);
            var select = document.getElementById('load_type');
            if (r.message.types) {
                var selector = "";
                for (i = 0; i < r.message.types.length; i++) {
                    if (r.message.types[i]['name'] === r.message.object_load_type) {
                        selector = "selected";
                    } else {
                        selector = "";
                    }
                    select.innerHTML += "<option value=\"" + r.message.types[i]['name'] 
                        + "\" " + selector + ">" + r.message.types[i]['name'] + "</option>";
                }
            }
        }
    });
}

function get_object_details(object, truck, customer, key) {
    // fetch object details
    frappe.call({
        'method': 'heimbohrtechnik.templates.pages.schlammanlieferung.get_object_details',
        'args': {
            'object_name': object,
            'truck': truck,
            'customer': customer,
            'key': key
        },
        'callback': function(r) {
            if (r.message.address) {
                var object_text = document.getElementById('object_text');
                object_text.innerHTML = r.message.address
            } else {
                console.log(r.message);
            }
        }
    });
}

function get_truck_weight(truck) {
    // fetch object details
    frappe.call({
        'method': 'heimbohrtechnik.templates.pages.schlammanlieferung.get_truck_weight',
        'args': {
            'truck': truck
        },
        'callback': function(r) {
            if (r.message.weight) {
                document.getElementById('empty_weight').value = r.message.weight;
                document.getElementById('empty_time').value = get_now();
                document.getElementById('empty_process_id').value = ":" + truck;
                document.getElementById('empty_scale').value = null;
            }
        }
    });
}

function compute_net() {
    var full_weight = document.getElementById('full_weight').value;
    var empty_weight = document.getElementById('empty_weight').value;
    if ((full_weight) && (empty_weight)) {
        var net_weight = full_weight - empty_weight;
        document.getElementById('net_weight').value = net_weight;
        if (net_weight > 0) {
            var btn_weigh = document.getElementById('weigh');
            btn_weigh.disabled = true;
            btn_weigh.style.visibility = "hidden";
            var btn_submit = document.getElementById('submit');
            btn_submit.disabled = false;
            btn_submit.style.visibility = "visible";
        }
    }
}

function get_now() {
    var now = new Date();
    var timestamp = now.getFullYear() + "-" + (now.getMonth() + 1) + "-" + now.getDate() + " " + now.getHours() + ":" + now.getMinutes() + ":" + now.getSeconds();
    return timestamp;
}

File: heimbohrtechnik/templates/pages/schlammanlieferung.py
# -*- coding: utf-8 -*-
# Copyright (c) 2021-2024, libracore and contributors
# License: AGPL v3. See LICENCE

from __future__ import unicode_literals
import frappe
from frappe import _
import json
import datetime
from frappe.utils import cint, flt

@frappe.whitelist(allow_guest=True)
def get_object_details(truck, customer, object_name, key):
    if validate_credentials(truck, customer, object_name, key):
        obj = frappe.get_doc("Object", object_name)
        return {
            'address': "{0}<br>{1}<br>{2}".format(obj.object_name, obj.object_street, obj.object_location)
        }
    else:
        return {'error': 'Not allowed'}

@frappe.whitelist(allow_guest=True)
def get_truck_weight(truck):
    try:
        weight = frappe.get_value("Truck", truck, "net_weight")
        return {'weight': weight}
    except Exception as err:
        return {'error': err}

@frappe.whitelist(allow_guest=True)
def get_default_scale():
    scale = frappe.get_value("MudEx Settings", "MudEx Settings", "default_scale")
    return scale
        
def validate_credentials(truck, customer, object_name, key):
    truck = frappe.get_doc("Truck", truck)
    if truck.customer == customer:
        object_key = frappe.get_value("Object", object_name, "object_key")
        if key == object_key:
            return True
        else:
            return False
    else:
        return False

@frappe.whitelist(allow_guest=True)
def insert_delivery(truck, customer, object_name, full_weight, empty_weight, net_weight, traces, load_type, ph, key):
    if not validate_credentials(truck, customer, object_name, key):
        return {'delivery': 'Invalid Key'}
        
    # prepare values
    allocation = None
    if frappe.db.exists("Object", object_name):
        o = frappe.get_doc("Object", object_name)
        allocation = [{
            'object': object_name,
            'oject_name': o.object_name,
            'object_street': o.object_street,
            'object_location': o.object_location,
            'weight': float(net_weight)
        }]
        if not load_type:
            load_type = o.load_type
    trace = json.loads(traces)
    for t in trace:
        t['weight'] = float(t['weight'])
    truck_doc = frappe.get_doc("Truck", truck)
    config = frappe.get_doc("MudEx Settings", "MudEx Settings")
    customer_doc = frappe.get_doc("Customer", customer)
    
    # create new record
    delivery = frappe.get_doc({
        'doctype': 'Truck Delivery',
        'truck': truck,
        'truck_description': truck_doc.title,
        'truck_owner': truck_doc.truck_owner,
        'truck_scale': config.default_scale,
        'date': datetime.datetime.now(),
        'customer': customer,
        'customer_name': customer_doc.customer_name,
        'full_weight': float(full_weight),
        'empty_weight': float(empty_weight),
        'net_weight': float(net_weight),
        'load_type': load_type,
        'ph': flt(ph),
        'objects': allocation,
        'trace': trace
    })
    delivery = delivery.insert(ignore_permissions=True)
    delivery.submit()
    return {'delivery': delivery.name}
    
@frappe.whitelist(allow_guest=True)
def get_load_types(object=None):
    all_load_types = frappe.get_all("Truck Load Type", fields=['name'])
    if object:
        load_type = frappe.get_value("Object", object, 'load_type')
    else:
        load_type = None
        
    return {'types': all_load_types, 'object_load_type': load_type}

@frappe.whitelist(allow_guest=True)
def set_ph(truck_delivery, truck, customer, object_name, key):
    if not validate_credentials(truck, customer, object_name, key):
        return "Invalid Key"
        
    # prepare values
    allocation = None
    if frappe.db.exists("Truck Delivery", truck_delivery):
        ph_sensor_name = frappe.get_value("MudEx Settings", "MudEx Settings", "default_ph_sensor")
        ph_sensor = frappe.get_doc("pH Sensor", ph_sensor_name)
        ph_sensor.read_sensor()
        if frappe.db.exists("Truck Delivery", truck_delivery):
            frappe.db.sql("""
                UPDATE `tabTruck Delivery`
                SET `ph` = %(ph)s
                WHERE `name` = %(name)s;
                """,
                {'ph': ph_sensor.ph, 'name': truck_delivery}
            )
            frappe.db.commit()
            return "Stored pH = {0} to {1}".format(ph_sensor.ph, truck_delivery)
            
        else:
            return "Invalid Truck Delivery"
    
    else:
        return "Truck delivery {0} not found".format(truck_delivery)
    
